### ISSUE TITLE
Customizable sitemap.xml

### ISSUE THREAD

**Issue by ugogo** [nozzle/react-static#986]
### Is this a feature request?
Yes.

---

Hey guys,

First of all, thanks a lot for this, I'm using `react-static` for a few projects and it works like a charm ❤️

Now, for **SEO** purposes I'd like to have a way to customize the generated sitemap.xml

According to Google, for multi-languages website

> You can use a Sitemap to tell Google all of the language and region variants for each URL. To do so, add a <loc> element specifying a single URL, with child <xhtml:link> entries listing every language/locale variant of the page including itself. Therefore if you have 3 versions of a page, your sitemap will have 3 entries, each with 3 identical child entries.

More details here ➡️https://support.google.com/webmasters/answer/189077?hl=en

tldr; I'd need a way to add a `<xhtml:link>` entry for each `<url>`

Also, it looks like the `<priority>` stuff is not useful anymore 🙊 

### Environment

1.  `react-static -V`: 5.9.12
2.  `node -v`: 8.11.2
3.  `npm -v`: 6.5.0

Cheers,

**Comment by lottamus** [comment]
I'd be happy to see a PR for this. Would you like to take a shot at it? 

Here's the v6 code: https://github.com/nozzle/react-static/blob/master/packages/react-static/src/static/buildXML.js

Here's the v5 code: https://github.com/nozzle/react-static/blob/v5/src/static/buildXML.js

**Comment by lottamus** [comment]
Is this a duplicate of https://github.com/nozzle/react-static/issues/964?

**Comment by tannerlinsley** [comment]
Not necessarily a duplicate, but very related. I would like to see the sitemap generation turned into a plugin, but this would be a breaking change, since we would have to pull it out of the default build and users would need to install it.

**Comment by marqpdx** [comment]
hey ugogo, what would your ideal xml look like for this? just paste a snippet that has the immediate parent element of `<url>` and everything within that `<url>` plus the new `<xhtml:link>` you are proposing.

hopefully it's straightforward to add and if so, i don't mind making a PR.

thx,
m

ps, why is the `<priority>` no longer needed?

**Comment by ugogo** [comment]
Hey @marqpdx (sorry @lottamus for not being more responsive, it has been some crazy days)

According to the Google documentation, it should be something like
```xml
<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
  xmlns:xhtml="http://www.w3.org/1999/xhtml">
  <url>
    <loc>http://www.example.com/english/page.html</loc>
    <xhtml:link 
               rel="alternate"
               hreflang="de"
               href="http://www.example.com/deutsch/page.html"/>
    <xhtml:link 
               rel="alternate"
               hreflang="de-ch"
               href="http://www.example.com/schweiz-deutsch/page.html"/>
    <xhtml:link 
               rel="alternate"
               hreflang="en"
               href="http://www.example.com/english/page.html"/>
  </url>
  <url>
    <loc>http://www.example.com/deutsch/page.html</loc>
    <xhtml:link 
               rel="alternate"
               hreflang="de"
               href="http://www.example.com/deutsch/page.html"/>
    <xhtml:link 
               rel="alternate"
               hreflang="de-ch"
               href="http://www.example.com/schweiz-deutsch/page.html"/>
    <xhtml:link 
               rel="alternate"
               hreflang="en"
               href="http://www.example.com/english/page.html"/>
  </url>
  <url>
    <loc>http://www.example.com/schweiz-deutsch/page.html</loc>
    <xhtml:link 
               rel="alternate"
               hreflang="de"
               href="http://www.example.com/deutsch/page.html"/>
    <xhtml:link 
               rel="alternate"
               hreflang="de-ch"
               href="http://www.example.com/schweiz-deutsch/page.html"/>
    <xhtml:link 
               rel="alternate"
               hreflang="en"
               href="http://www.example.com/english/page.html"/>
  </url>
</urlset>
```
So from what I understand, for each declared `<loc>`, you should have a `<xhtml:link>` for each language in your application.
As `react-static` does not support multiple languages by default, being able to update the `buildXML` step through the main configuration file seems to be the way to go, WDYT?

About the `<priority>`, still according to Google (https://support.google.com/webmasters/answer/183668?hl=en)

> Google supports several sitemap formats, described here. Google expects the standard sitemap protocol in all formats. Google does not currently consume the <priority> attribute in sitemaps.

**Comment by marqpdx** [comment]
hi ugogo

thanks for this.

it seems like to do this properly, we'll have to have a way to store configured data about the languages in use for site, and then a way to have multiple pages per route, per language. 

like you mentioned, this could go into a main config file, but i think it would be good to hear from tannerlinsley and the crew about how multiple languages will be handled, and then we can update the sitemap.xml functionality to match.

thanks all,
m

**Comment by ugogo** [comment]
Hey guys @marqpdx @tanayseven 👋 

IMO the languages stuff is part of a larger project.
Would you be fine if we add an option into the main config file?
It could loop into each route and returns a new `String`

**Comment by tannerlinsley** [comment]
The v7 branch has the sitemap generation extracted as a plugin using the `afterExport` plug.  Might be worth looking at for ideas or inspiration, but this where I see things going.

**Comment by tannerlinsley** [comment]
The new sitemap plugin in v7 should offer a lot more customization on all fronts. Be sure to check it out!
